suse/distributionscanner.go

```diff
@@ -111,11 +88,46 @@ func (ds *DistributionScanner) Scan(ctx context.Context, l *claircore.Layer) ([]
 // distribution if it exists.
 //
 // separated into its own method to aid testing.
-func (ds *DistributionScanner) parse(buff *bytes.Buffer) *claircore.Distribution {
-	for _, ur := range suseRegexes {
-		if ur.regexp.Match(buff.Bytes()) {
-			return releaseToDist(ur.release)
+func (ds *DistributionScanner) parse(ctx context.Context, buff *bytes.Buffer) *claircore.Distribution {
+	kv, err := osrelease.Parse(ctx, buff)
+	if err != nil {
+		zlog.Warn(ctx).Err(err).Msg("malformed os-release file")
+	}
+	cpeName, cpeOK := kv["CPE_NAME"]
+	if !cpeOK {
+		return nil
+	}
+	// Instead of regexing through, we can grab the CPE.
+	c, err := cpe.Unbind(cpeName)
+	if err != nil {
+		zlog.Warn(ctx).Err(err).Msg("could not unbind CPE")
+		return nil
+	}
+
+	d, err := cpeToDist(c)
+	if err != nil {
+		zlog.Warn(ctx).Err(err).Msg("error converting cpe to distribution")
+		return nil
+	}
+
+	return d
+}
+
+func cpeToDist(r cpe.WFN) (*claircore.Distribution, error) {
+	if vendor, err := cpe.NewValue("opensuse"); err == nil && r.Attr[cpe.Vendor] == vendor {
```

Comment:
nit: can make the CPEs globals so they're only evaluated once. Haven't looked into the function to see if it saves a ton of work or not, though

Reply:
My instinct is that any minimal efficiency gain would be lost in legibility:

```go
var (
	openSUSECPEVal, leapCPEVal, suseCPEVal, slesCPEVal cpe.Value
)

func init() {
	var err error
	openSUSECPEVal, err = cpe.NewValue("opensuse")
	if err != nil {
		panic(err)
	}
	leapCPEVal, err = cpe.NewValue("leap")
	if err != nil {
		panic(err)
	}
	suseCPEVal, err = cpe.NewValue("suse")
	if err != nil {
		panic(err)
	}
	slesCPEVal, err = cpe.NewValue("sles")
	if err != nil {
		panic(err)
	}
}

func cpeToDist(r cpe.WFN) (*claircore.Distribution, error) {
	if r.Attr[cpe.Vendor] == openSUSECPEVal {
		if r.Attr[cpe.Product] == leapCPEVal {
			return mkLeapDist(r.String(), r.Attr[cpe.Version].String()), nil
		}
	}
...
```
what do you think?

Reply:
fine as-is 👍